src/Cms/Binding.php

```diff
@@ -0,0 +1,417 @@
+		// `site://filename.jpg` or `user://parent-uuid/filename.jpg`,
+		// thus we need to resolve the parent UUID to its model
+		// and then query for the file by filename
+		if ($this->type() === 'file') {
+			$parent   = Binding::for($this->protocol->base())->model();
+			$filename = $this->protocol->path()->toString();
+			return $parent->file($filename);
+		}
+
+		// TODO: resolving values for blocks and structure
+		// schema://$parent→uuid()/$fieldName/$block→id()
+		// 'block' => Binding::for($firstPart)->model()->$fieldPart()->toBlocks()->get($lastPart),
+
+		// schema://$parent→uuid()/$fieldName/$structureItem→id()
+		// 'block' => Binding::for($firstPart)->model()->$fieldPart()->toStructure()->findBy('uuid', $lastPart)
+	}
+
+	/**
+	 * Lookup page/file by traversing through index
+	 * and finding the model with the matching UUID hash
+	 * in its content file
+	 *
+	 * @param bool $global
+	 * @return \Kirby\Cms\Page|\Kirby\Cms\File|null
+	 */
+	protected function findFromIndex(bool $global = false): Page|File|null
+	{
+		// get index collection - either with current collection
+		// as context or with global index
+		$index = $this->collection($global);
+
+		// TODO: this would likely work differently for block and structure
+
+		// Note: we cannot filter by `uuid` as that would
+		// call the method that creates a binding instance and
+		// generates a UUID for each model. We only actually
+		// want to check for models that already have a UUID
+		// written to the content file
+		$id = $this->id();
+		return $index->filter(
+			fn ($model) => $model->content()->get('uuid')->value() === $id
+		)->first();
+	}
+
+
+	/**
+	 * Shorthand to create a binding by passing either
+	 * an UUID or a site/page/file/user instance
+	 *
+	 * @param string|\Kirby\Cms\Page|\Kirby\Cms\File|\Kirby\Cms\User|\Kirby\Cms\Site $base
+	 * @param \Kirby\Cms\Models|null $collection
+	 * @return static
+	 */
+	public static function for(
+		string|Page|File|User|Site $seed,
+		Models|null $collection = null
+	): static {
+		// UUID
+		if (is_string($seed) === true) {
+			return new static(uuid: $seed, collection: $collection);
+		}
+
+		// site/page/file/user instance
+		return new static(model: $seed, collection: $collection);
+	}
+
+	/**
+	 * Return (if needed, generate) UUID hash
+	 *
+	 * @return string
+	 */
+	public function id(): string
+	{
+		// if no ID hash (protocol's host) exists yet,
+		// generate a new one and update model as well as protocol
+		// (should only be entered for pages or files)
+		if ($this->protocol->host() === null) {
+			$this->create();
+		}
+
+		return $this->protocol->tostring(false);
+	}
+
+	/**
+	 * Populates cache with UUIDs
+	 * for all pages and files in index
+	 *
+	 * @return void
+	 */
+	public static function index(): void
+	{
+		$kirby = App::instance();
+		$site  = $kirby->site();
+		$index = $site->index(true);
+
+		// pages
+		foreach ($index as $page) {
+			static::for($page)->populate();
+		}
+
+		// page files
```

Comment:
I'm not sure if it makes a huge difference, but it feels better to me to do this in a nested foreach loop of the page loop above.